src/Silverback.Integration/Messaging/Validation/MessageValidator.cs

```diff
@@ -0,0 +1,51 @@
+// Copyright (c) 2020 Sergio Aquilini
+// This code is licensed under MIT license (see LICENSE file for details)
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Linq;
+
+namespace Silverback.Messaging.Validation
+{
+    /// <summary>
+    ///     Validates the message and throws a <see cref="MessageValidationException" /> depending on the
+    ///     <see cref="MessageValidationMode" />.
+    /// </summary>
+    internal static class MessageValidator
+    {
+        /// <summary>
+        ///     Validates the message.
+        /// </summary>
+        /// <param name="message">The message.</param>
+        /// <param name="validationMode">The message validation mode.</param>
+        /// <param name="callback">The callback function to log the message validation errors.</param>
+        /// <exception cref="MessageValidationException">The thrown exception when validation mode is ThrowException.</exception>
+        public static void ValidateMessage(
+            object message,
+            MessageValidationMode validationMode,
+            Action<string> callback)
+        {
+            var validationContext = new ValidationContext(message);
+            var results = new List<ValidationResult>();
+            var validMessage = Validator.TryValidateObject(message, validationContext, results, true);
+
+            if (validMessage)
+            {
+                return;
+            }
+
+            if (validationMode == MessageValidationMode.ThrowException)
+            {
+                throw new
+                    MessageValidationException(
+                        $"Message not valid. {string.Join(Environment.NewLine, results.Select(validationResult => validationResult.ErrorMessage))}");
+            }
+
+            callback(
+                string.Join(
```

Comment:
This `string.Jojn(...)` is duplicated with the one done directly in the interpolation at line 42. Could be refactored.